compiler/types.nim

```diff
@@ -1144,6 +1145,13 @@ proc sameTypeAux(x, y: PType, c: var TSameTypeClosure): bool =
   of tyEmpty, tyChar, tyBool, tyNil, tyPointer, tyString, tyCstring,
      tyInt..tyUInt64, tyTyped, tyUntyped, tyVoid:
     result = sameFlags(a, b)
+    if result and PickyCAliases in c.flags:
+      # additional requirement for the caching of generics for importc'ed types:
+      # the symbols must be identical too:
```

Comment:
Given that I should learn: Why is checking the flags here equivalent to the symbols being identical? Is it just because we have checked for equality of the actual type etc. before in this proc?

Also it seems confusing that `sameFlags` is called before, but then we check more flags here.

In general what makes reading compiler code hard for me is that most procs have no short comment about their intent, e.g. this `sameTypeAux` could use a couple of lines saying what is needed for a type to be considered the same for instance.

Reply:
> Given that I should learn: Why is checking the flags here equivalent to the symbols being identical? Is it just because we have checked for equality of the actual type etc. before in this proc?

Because we have established they are of the type *kind* which for `tyInt` etc means they are atomic types. Arguably we could also have checked for `a.sym == b.sym` instead but this would be worse for type aliases to `cint`, e.g. `type myint = cint`.

> Also it seems confusing that sameFlags is called before, but then we check more flags here.

`sameFlags` looks at the type flags but `importc` is stored as a sym flag. These are different things.

> this sameTypeAux could use a couple of lines saying what is needed for a type to be considered the same for instance.

Nim's type system has both structural and nominal type system aspect and type equality is supposed to follow from the spec.

Reply:
Thanks for the clarification!